tortoisebeacon/tortoise_beacon.go

```
@@ -193,76 +203,178 @@ func (tb *TortoiseBeacon) Close() {
+		log.Uint64("weight", mostWeight))
+	return beacon
+}
+
+// GetBeacon returns a Tortoise Beacon value as []byte for a certain epoch or an error if it doesn't exist.
+// TODO(nkryuchkov): consider not using (using DB instead)
+func (tb *TortoiseBeacon) GetBeacon(epochID types.EpochID) ([]byte, error) {
+	if epochID == 0 {
+		return nil, ErrZeroEpoch
+	}
 
-		return nil, ErrBeaconNotCalculated
+	beaconEpoch := epochID - 1
+	if beaconEpoch.IsGenesis() {
+		return types.HexToHash32(genesisBeacon).Bytes(), nil
 	}
 
-	return beacon.Bytes(), nil
+	beacon := tb.getBeacon(beaconEpoch)
+	if beacon != nil {
+		return beacon, nil
+	}
+
+	beacon = tb.getPersistedBeacon(beaconEpoch)
+	if beacon != nil {
+		return beacon, nil
+	}
+
+	tb.logger.With().Warning("beacon not available",
+		epochID,
+		log.Uint32("beacon_epoch", uint32(beaconEpoch)))
+	return nil, ErrBeaconNotCalculated
+}
+
+func (tb *TortoiseBeacon) getBeacon(epoch types.EpochID) []byte {
+	tb.mu.RLock()
+	defer tb.mu.RUnlock()
+	if beacon, ok := tb.beacons[epoch]; ok {
+		return beacon.Bytes()
+	}
+	return nil
 }
 
 func (tb *TortoiseBeacon) setBeacon(epoch types.EpochID, beacon types.Hash32) {
+	tb.persistBeacon(epoch, beacon)
 	tb.mu.Lock()
+	defer tb.mu.Unlock()
 	tb.beacons[epoch] = beacon
-	tb.mu.Unlock()
+}
+
+func (tb *TortoiseBeacon) persistBeacon(epoch types.EpochID, beacon types.Hash32) {
+	if tb.db == nil {
+		return
+	}
+	if err := tb.db.Put(epoch.ToBytes(), beacon.Bytes()); err != nil {
+		// persisting beacon is only best-effort basis.
+		// if the node goes offline and back online in the middle of the same epoch, it simply waits
+		// enough blocks that reports beacon, or till the next epoch.
+		tb.logger.With().Error("failed to persist beacon",
+			epoch,
+			log.String("beacon", beacon.ShortString()), log.Err(err))
+	}
+}
+
+func (tb *TortoiseBeacon) getPersistedBeacon(epoch types.EpochID) []byte {
+	if tb.db == nil {
+		return nil
+	}
+	if beacon, err := tb.db.Get(epoch.ToBytes()); err == nil {
```

why not to return an error?

---

done